openmdao/core/system.py

```diff
@@ -715,6 +717,612 @@ def _declare_options(self):
+            consistent with the shape arg (if given), or (3) an array_like matching the shape of
+            val, if val is array_like. A value of None means this output has no lower bound.
+            Default is None.
+        upper : float or list or tuple or ndarray or None
+            Upper bound(s) in user-defined units. It can be (1) a float, (2) an array_like
+            consistent with the shape arg (if given), or (3) an array_like matching the shape of
+            val, if val is array_like. A value of None means this output has no upper bound.
+            Default is None.
+        ref : float
+            Scaling parameter. The value in the user-defined units of this output variable when
+            the scaled value is 1. Default is 1.
+        ref0 : float
+            Scaling parameter. The value in the user-defined units of this output variable when
+            the scaled value is 0. Default is 0.
+        res_ref : float
+            Scaling parameter. The value in the user-defined res_units of this output's residual
+            when the scaled value is 1. Default is None, which means residual scaling matches
+            output scaling.
+        """
+        # Cache the solver options for use later in the setup process.
+        # Since this can be called before setup, there is no way to update the
+        # self._var_allprocs_abs2meta['output'] values since those have not been setup yet.
+        # These values are applied in the System._apply_output_solver_options method
+        # which is called in System._setup. That method is only called by the top model.
+
+        output_solver_options = {}
+        if lower is not _UNDEFINED:
+            output_solver_options['lower'] = lower
+        if upper is not _UNDEFINED:
+            output_solver_options['upper'] = upper
+        if ref is not _UNDEFINED:
+            output_solver_options['ref'] = ref
+        if ref0 is not _UNDEFINED:
+            output_solver_options['ref0'] = ref0
+        if res_ref is not _UNDEFINED:
+            output_solver_options['res_ref'] = res_ref
+        self._output_solver_options[name] = output_solver_options
+        return
+
+    def _apply_output_solver_options(self):
+        """
+        Apply the cached output solver options.
+
+        Solver options can be set using the System.set_output_solver_options method.
+        These cannot be set immediately when that method is called because not
+        all the variables have been setup at the time a user could potentially want to call it.
+        So they are cached so that they can be applied later in the setup process.
+        They are applied in System._setup using this method.
+        """
+        # Loop through the output solver options that have been set on this System
+        for name, options in self._output_solver_options.items():
+            from openmdao.core.group import Group
```

Comment:
Should move the import outside of the loop.  Could also eliminate the import and the isinstance check and do something like
```
subsys_path = name.rpartition('.')[0]
subsys = self._get_subsystem(subsys_path) if subsys_path else self
```

Reply:
Used your code.